### PR TITLE
CORE-9550 Add full public key id for signing keys in crypto worker

### PR DESCRIPTION
**Pull request by kyriathar**
The reason for this change is potential clashes in short key ids during keys look up, which could lead in returning a key which is not owned by the caller.

- adds full key id in `crypto_signing_key` DB table.
- introduces avro types for short and full key ids (`ShortHashes` for short key ids and `SecureHashes` for full key ids)

runtime-os PR: [#3091](https://github.com/corda/corda-runtime-os/pull/3091)

### CODE REVIEW

File: data/avro-schema/src/main/resources/avro/net/corda/data/crypto/wire/ops/rpc/queries/ByIdsRpcQuery.avsc
Comment (mnesbit) [inline data/avro-schema/src/main/resources/avro/net/corda/data/crypto/wire/ops/rpc/queries/ByIdsRpcQuery.avsc:5]: why the limit of 20? Does the backend code break up large requests into a series of size 20 batch requests?

File: data/avro-schema/src/main/resources/avro/net/corda/data/crypto/ShortHashes.avsc
Comment (dickon) [inline data/avro-schema/src/main/resources/avro/net/corda/data/crypto/ShortHashes.avsc:3]: Is this temporary code or do we want to keep both kinds of hash long term, and if so, why do we need 2?

File: data/avro-schema/src/main/resources/avro/net/corda/data/crypto/wire/ops/rpc/queries/ByIdsRpcQuery.avsc
Comment (kyriathar) [reply data/avro-schema/src/main/resources/avro/net/corda/data/crypto/wire/ops/rpc/queries/ByIdsRpcQuery.avsc:5]: `SigningKeyStoreImpl` look up methods throw if requested keys found to be in number more than 20 (this was already there).

File: data/avro-schema/src/main/resources/avro/net/corda/data/crypto/ShortHashes.avsc
Comment (dickon) [reply data/avro-schema/src/main/resources/avro/net/corda/data/crypto/ShortHashes.avsc:3]: We will keep both hash formats, so that short forms can be used in HTTP address query strings.

File: data/avro-schema/src/main/resources/avro/net/corda/data/crypto/wire/ops/rpc/queries/ByIdsRpcQuery.avsc
Comment (dickon) [inline data/avro-schema/src/main/resources/avro/net/corda/data/crypto/wire/ops/rpc/queries/ByIdsRpcQuery.avsc]: whitespace looks wrong - the closing ] should be algiedn with the open double quote of "type"

File: data/avro-schema/src/main/resources/avro/net/corda/data/crypto/ShortHashes.avsc
Comment (dickon) [inline data/avro-schema/src/main/resources/avro/net/corda/data/crypto/ShortHashes.avsc]: better to say something like "List of short hashes. A short hash is the first 12 characters of the hex form of the hash. Short hashes should only be used when conciseness is important."

File: data/avro-schema/src/main/resources/avro/net/corda/data/crypto/wire/ops/rpc/queries/ByIdsRpcQuery.avsc
Comment (dickon) [reply data/avro-schema/src/main/resources/avro/net/corda/data/crypto/wire/ops/rpc/queries/ByIdsRpcQuery.avsc:5]: This number 20 (and the version inByIdsFlowQuery) feels like a server limitation rather than something which should be in the interface.

But it's already there, we should highlight in both places as you are doing if we we are to mention it one place, and I wouldn't hold up approval on this.

**Review by dickon — CHANGES_REQUESTED** [review]
LGTM, other than some very minor nitpicks. I should be able to quickly approve once you've done the rename you want to do and addressed the nitpicks.

File: data/avro-schema/src/main/resources/avro/net/corda/data/crypto/ShortHashes.avsc
Comment (dickon) [reply data/avro-schema/src/main/resources/avro/net/corda/data/crypto/ShortHashes.avsc]: actually, now I've looked at the implementation and commented there, better to simply allow unique prefixes, in which case either:

a) no AVRO change at all - anywhere you can specify a hash you can use a unique prefix
b) we have a more general AVRO change which allows a unique hash prefix to be sent, in a way that can be differentiated from a full hash.

**Review by dickon — CHANGES_REQUESTED** [review]
Now I understand the database level, my thinking is that an arbitrary fixed prefix length is more trouble than arbitary unique prefix length, so changing review status.

File: data/avro-schema/src/main/resources/avro/net/corda/data/crypto/wire/ops/rpc/queries/ByIdsRpcQuery.avsc
Comment (kyriathar) [reply data/avro-schema/src/main/resources/avro/net/corda/data/crypto/wire/ops/rpc/queries/ByIdsRpcQuery.avsc]: Addressed

File: data/avro-schema/src/main/resources/avro/net/corda/data/crypto/ShortHashes.avsc
Comment (kyriathar) [reply data/avro-schema/src/main/resources/avro/net/corda/data/crypto/ShortHashes.avsc]: Doc updated with suggestion

**Review by mnesbit — DISMISSED** [review]
OK

File: data/avro-schema/src/main/resources/avro/net/corda/data/crypto/wire/ops/flow/queries/ByIdsFlowQuery.avsc
Comment (dickon) [inline data/avro-schema/src/main/resources/avro/net/corda/data/crypto/wire/ops/flow/queries/ByIdsFlowQuery.avsc]: We should not allow 48 bit ShortHashes for this query; it'd make it pointlessly easier for flows to pick up keypairs they shouldn't.

**Review by dickon — CHANGES_REQUESTED** [review]

File: data/avro-schema/src/main/resources/avro/net/corda/data/crypto/wire/ops/flow/queries/ByIdsFlowQuery.avsc
Comment (kyriathar) [reply data/avro-schema/src/main/resources/avro/net/corda/data/crypto/wire/ops/flow/queries/ByIdsFlowQuery.avsc]: Removed short key ids from `ByIdsFlowQuery`.

File: data/avro-schema/src/main/resources/avro/net/corda/data/crypto/ShortHashes.avsc
Comment (dickon) [inline data/avro-schema/src/main/resources/avro/net/corda/data/crypto/ShortHashes.avsc]: Please make this something like:

  List of short hashes. A short hash is the first 12 characters of the hex string form of the secure hash, and therefore has only 48 bits of entropy. Short hashes must only be used when arrangements have been made to prevent or be safe if there are collisions, whether accidental or malicious. Only use this where necessary, for instance due to the space required for a full size secure hash being prohibitive.

**Review by dickon — CHANGES_REQUESTED** [review]
LGTM, other than we need to clarify the collision risk with short hashes.

File: data/avro-schema/src/main/resources/avro/net/corda/data/crypto/ShortHashes.avsc
Comment (dickon) [inline data/avro-schema/src/main/resources/avro/net/corda/data/crypto/ShortHashes.avsc]: Should end in a full stop

**Review by dickon — DISMISSED** [review]
LGTM, other than a small punctuation problem in the docstring which is not blocking.

File: data/avro-schema/src/main/resources/avro/net/corda/data/crypto/ShortHashes.avsc
Comment (kyriathar) [reply data/avro-schema/src/main/resources/avro/net/corda/data/crypto/ShortHashes.avsc]: Done

**Review by dickon — DISMISSED** [review]
LGTM

**Review by mnesbit — APPROVED** [review]
OK

**Pull request merged**
